### ISSUE TITLE
Issue: Gradient Vanishing with `attorch.nn.BatchNorm2d`

### ISSUE THREAD

**Issue by 2niuhe** [BobMcDear/attorch#8]
I encountered an issue while using `attorch.nn.BatchNorm2d`. After performing a backward pass, I noticed that the gradients are vanishing, with most elements in `input.grad` being zero. In contrast, the gradient computation for `torch.nn.BatchNorm2d` appears to be functioning correctly.

#### Steps to Reproduce:

```python
import torch
import attorch
from attorch.nn import BatchNorm2d

shape = (2, 4, 2, 2)
bn = BatchNorm2d(4, device='cuda')
inp = torch.randn(shape, requires_grad=True, device='cuda')

# Forward pass
output = bn(inp)

# Backward pass
output.sum().backward()

print(inp.grad)
```

#### Expected Behavior:
The gradients in `input.grad` should not be  zero.

#### Actual Behavior:
Most elements in `input.grad` are zero after the backward pass.

input shape is (4,32,16,16)

- For `attorch.nn.BatchNorm2d`:
  - Number of zeros in `input.grad`: 32,450
  - Number of non-zeros in `input.grad`: 318

- For `torch.nn.BatchNorm2d`:
  - Number of zeros in `input.grad`: 0
  - Number of non-zeros in `input.grad`: 32,768

Despite `torch` calculating a maximum gradient value of only `tensor(6.3541e-08)`, it still produces valid gradients.

Thank you for your assistance!

**Comment by BobMcDear** [comment]
Hello,

This shouldn't pose a problem in practice because the network incoprorates layers besides batch norm that'll result in much larger gradients. Below is an example that illustrates my point; if you examine ```attorch_grads```, you'd observe that the gradients are not vanishing, as is the case with ```pytorch_grads```. 

```python
from collections import defaultdict

import attorch
import torch
from torch import nn
from torch.nn import functional as F
from torch.optim import AdamW


def create_model(bn):
    torch.manual_seed(0)
    return nn.Sequential(nn.Conv2d(n_in, n_hidden, kernel_size=3, stride=2),
                         bn(n_hidden),
                         nn.ReLU(),
                         nn.Conv2d(n_hidden, n_hidden, kernel_size=3, stride=2),
                         bn(n_hidden),
                         nn.ReLU(),
                         nn.AdaptiveAvgPool2d(1),
                         nn.Conv2d(n_hidden, n_in, kernel_size=1))


def train(model):
    torch.manual_seed(0)
    model = model.to('cuda')
    input = torch.randn(batch_size, n_in, l, l, device='cuda')
    target = F.adaptive_avg_pool2d(F.relu(input ** 2 + torch.sin(input)), 1) # Contrived target
    opt = AdamW(model.parameters(), 3e-4)

    for _ in range(100):
        loss = F.mse_loss(model(input), target)
        loss.backward()
        opt.step()
        opt.zero_grad()


def get_gradients(model):
    def save_gradients(module, grad_input, _):
        gradients[module].append(grad_input)

    gradients = defaultdict(list)
    model[1].register_full_backward_hook(save_gradients)
    model[4].register_full_backward_hook(save_gradients)

    train(model)
    return gradients


batch_size = 32
n_in = 3
l = 64
n_hidden = 64

# Gradients per batch norm module for each training iteration
pytorch_grads = get_gradients(create_model(nn.BatchNorm2d))
attorch_grads = get_gradients(create_model(attorch.BatchNorm2d))
```

For a real-world use case, please refer to [the Imagenette example](https://github.com/BobMcDear/attorch/tree/main/examples/imagenette), where an attorch-backed ResNet with batch norm will converge to the same accuracy as the PyTorch baseline.

**Comment by 2niuhe** [comment]
Thank you for your response and the detailed explanation. 

Your attorch library has been a significant help in my learning of Triton. Thank you once again for your assistance!
